File: main/count-luck/count-luck-iter.py
#!/usr/bin/env python3

import contextlib


def read_value():
    """Reads a line as a single integer."""
    return int(input())


def read_record():
    """Reads a line as a record of integers."""
    return map(int, input().split())


def read_row(width):
    """Reads a grid row of the specific width."""
    row = list(input().strip())
    if len(row) != width:
        raise ValueError('wrong row width')
    return row


def read_grid():
    """Reads a height value, width value, and rows as a grid."""
    height, width = read_record()
    return [read_row(width) for _ in range(height)]


def height(grid):
    """Gets the height of a rectangular grid (stored row-major)."""
    return len(grid)


def width(grid):
    """Gets the width of a rectangular grid (stored row-major)."""
    try:
        return len(grid[0])
    except IndexError:
        return 0


def find_start(grid):
    """Finds the coordinates of the start mark "M" in the grid."""
    for i, row in enumerate(grid):
        with contextlib.suppress(ValueError):
            return (i, row.index('M'))

    raise ValueError('start mark "M" not found in grid')


def can_go(grid, i, j):
    """Checks if the cell (i, j) is on the grid and accessible."""
    return 0 <= i < height(grid) and 0 <= j < width(grid) and grid[i][j] != 'X'


def neighbors(grid, i, j):
    """Gets the coordinates of accessible adjacent cells to (i, j) on grid."""
    possible_neighbors = ((i - 1, j), (i + 1, j), (i, j - 1), (i, j + 1))
    return [(h, k) for h, k in possible_neighbors if can_go(grid, h, k)]


class Frame:
    """Stack frame for traversing the grid using backtracking."""

    def __init__(self, junctions, grid, i, j):
        """Makes a frame for (i, j) on grid with a junction count so far."""
        nei_list = neighbors(grid, i, j)
        self.junctions = (junctions + 1 if len(nei_list) > 1 else junctions)
        self.nei_iter = iter(nei_list)


def count_junctions(grid, i, j):
    """
    Counts the number of junctions from (i, j) to the finish("*").
    Mutates grid, filling at least some of it in with "X"es.
    """
    grid[i][j] = 'X'
    stack = [Frame(0, grid, i, j)]

    while stack:
        frame = stack[-1]
        try:
            i, j = next(frame.nei_iter)
            if grid[i][j] == '*':
                return frame.junctions
            grid[i][j] = 'X'
            stack.append(Frame(frame.junctions, grid, i, j))
        except StopIteration:
            del stack[-1]

    raise ValueError('could not find a path from start to finish')


def run():
    """
    Reads a grid and guess, counts junctions from start to finish,
    and reports if the guess was correct.
    """
    grid = read_grid()
    guess = read_value()
    junctions = count_junctions(grid, *find_start(grid))
    print('Impressed' if guess == junctions else 'Oops!')


if __name__ == '__main__':
    for _ in range(read_value()):
        run()

File: main/count-luck/count-luck-iter.cpp
#include <algorithm>
#include <array>
#include <cstdlib>
#include <iostream>
#include <iterator>
#include <limits>
#include <stack>
#include <string>
#include <tuple>
#include <vector>

namespace {
    using Grid = std::vector<std::string>;

    using Pos = std::tuple<std::size_t, std::size_t>;

    constexpr std::size_t sides {4u};

    constexpr auto trapped = -1;

    Grid read_grid()
    {
        std::size_t height {}, width {};
        std::cin >> height >> width;
        Grid grid (height);

        for (auto& row : grid) {
            std::cin >> row;
            if (row.size() != width) std::abort();
        }

        return grid;
    }

    Pos find_start(const Grid& grid) noexcept
    {
        for (std::size_t i {0u}; i != grid.size(); ++i) {
            for (std::size_t j {0u}; j != grid[i].size(); ++j)
                if (grid[i][j] == 'M') return std::make_tuple(i, j);
        }

        std::abort(); // Crash "nicely" if there is no start location.
    }

    struct Frame {
        Pos pos;
        std::size_t index;
        int junctions;
    };

    constexpr std::array<Pos, sides>
    neighbors(const std::size_t i, const std::size_t j) noexcept
    {
        return {{{i - 1u, j}, {i + 1u, j}, {i, j - 1u}, {i, j + 1u}}};
    }

    inline bool can_go(const Grid& grid,
                       const std::size_t i, const std::size_t j) noexcept
    {
        return i < grid.size() && j < grid[i].size() && grid[i][j] != 'X';
    }

    inline int count_choices(const Grid& grid,
                             const std::size_t i, const std::size_t j) noexcept
    {
        auto choices = 0;

        for (const auto& cell : neighbors(i, j)) {
            std::size_t h {}, k {};
            std::tie(h, k) = cell;
            if (can_go(grid, h, k)) ++choices;
        }

        return choices;
    }

    int count_junctions(Grid& grid, std::size_t i, std::size_t j)
    {
        std::stack<Frame> frames;
        frames.push({{i, j}, 0u, 0});

        while (!frames.empty()) {
            auto& frame = frames.top();

            if (frame.index == sides) {
                frames.pop();
                continue;
            }

            std::tie(i, j) = frame.pos;

            if (frame.index == 0) {
                if (grid[i][j] == '*') return frame.junctions;
                grid[i][j] = 'X';

                const auto choices = count_choices(grid, i, j);

                if (choices == 0) {
                    frames.pop();
                    continue;
                }

                if (choices > 1) ++frame.junctions;
            }

            std::size_t h {}, k {};
            std::tie(h, k) = neighbors(i, j)[frame.index++];
            if (can_go(grid, h, k)) frames.push({{h, k}, 0u, frame.junctions});
        }

        return trapped;
    }
}

int main()
{
    int t {};
    for (std::cin >> t; t > 0; --t) {
        auto grid = read_grid();

        int guess {};
        std::cin >> guess;

        const auto start = find_start(grid);
        std::size_t i {}, j {};
        std::tie(i, j) = start;

        const auto count = count_junctions(grid, i, j);
        std::cout << (count == guess ? "Impressed" : "Oops!") << '\n';
    }
}
